drivers: display: st7789v: Move params to `struct config`

caspermeijn opened the pull request:
Move the const configuration parameters to `struct st7789v_config`.

Signed-off-by: Casper Meijn <casper@meijn.net>

caspermeijn committed:
drivers: display: st7789v: Move params to `struct config`

Move the const configuration parameters to `struct st7789v_config`.

Signed-off-by: Casper Meijn <casper@meijn.net>

jfischer-no commented on drivers/display/display_st7789v.c:
```diff
@@ -288,54 +303,58 @@ static void st7789v_lcd_init(const struct device *dev)
 
-	st7789v_transmit(dev, ST7789V_CMD_PVGAMCTRL, st7789v_pvgam_param,
-			 sizeof(st7789v_pvgam_param));
+	memcpy(tmp_buf, config->pvgam_param, sizeof(config->pvgam_param));
```
Do not bother with "warning: passing argument 3 of 'blabla' discards 'const'", just:
`st7789v_transmit(dev, ST7789V_CMD_PVGAMCTRL, (void *)config->pvgam_param, sizeof(config->pvgam_param));`

caspermeijn replied:
I added the const conversion as requested.

jfischer-no commented on drivers/display/display_st7789v.c:
```diff
@@ -437,15 +448,34 @@ static const struct display_driver_api st7789v_api = {
 DEVICE_DT_INST_DEFINE(0, &st7789v_init,
-	      PM_DEVICE_DT_INST_REF(0), &st7789v_data, NULL, POST_KERNEL,
-	      CONFIG_DISPLAY_INIT_PRIORITY, &st7789v_api);
+		      PM_DEVICE_DT_INST_REF(0), &st7789v_data, &st7789v_config, POST_KERNEL,
```
It is already wrapped, so please wrap it again, here and similar places, so that where it is not necessary does not exceed the 80 characters width.

caspermeijn replied:
Done. I did already run uncrustify to make the code look better ;-)

caspermeijn commented:
Rebased on main, where `struct st7789v_config` already exists.

gmarull commented on drivers/display/display_st7789v.c:
```diff
@@ -409,11 +424,27 @@ static const struct display_driver_api st7789v_api = {
 	.bus = SPI_DT_SPEC_INST_GET(0, SPI_OP_MODE_MASTER | SPI_WORD_SET(8), 0),
+	.vcom = DT_INST_PROP(0, vcom),
+	.gctrl = DT_INST_PROP(0, gctrl),
+	.vdv_vrh_enable = (DT_INST_NODE_HAS_PROP(0, vrhs) && DT_INST_NODE_HAS_PROP(0, vdvs)),
+	.vrh_value = DT_INST_PROP(0, vrhs),
+	.vdv_value = DT_INST_PROP(0, vdvs),
```
how does this work if `vrhs` or `vdvs` are not present? Are these boolean props?

caspermeijn replied:
Good catch! These are integer properties.  I tested that the build fails if one of these is missing. I added a default value to fix it.

gmarull reviewed (commented):
Can driver be made multi-instance, too?

caspermeijn commented:
> Can driver be made multi-instance, too?

I prefer to finished #43251 before starting multi-instance. There is still an `#if` in the config struct for a gpio pin. This could be handled by a optional initialization of the config data, but that requires the initialization fix in #43251.

carlescufi commented:
@caspermeijn please rebase

caspermeijn committed:
drivers: display: st7789v: Remove compile time reset selection

Replace the compiletime `#if` with a runtime decision. This makes
a multi-instance driver possible.

Signed-off-by: Casper Meijn <casper@meijn.net>

caspermeijn committed:
drivers: display: st7789v: Add multi-instance support

Replace the config and data struct with a macro. Execute the macro
for each instance.

Signed-off-by: Casper Meijn <casper@meijn.net>

caspermeijn commented:
> > Can driver be made multi-instance, too?
> 
> I prefer to finished #43251 before starting multi-instance. There is still an `#if` in the config struct for a gpio pin. This could be handled by a optional initialization of the config data, but that requires the initialization fix in #43251.

@gmarull In the meantime the initialization fix is merged, so I added multi-instance support.

@carlescufi rebased
